tests/chainer_tests/link_hooks_tests/test_spectral_normalization.py

```diff
@@ -176,23 +159,120 @@ def check_u_not_updated_in_test(self, gpu):
+        testing.assert_allclose(y1, y2, **self.allclose_options)
 
-    def test_u_not_updated_in_test_cpu(self):
+    def test_u_not_updated_in_test(self, backend_config):
         if not self.lazy_init:
-            self.check_u_not_updated_in_test(False)
+            self.check_u_not_updated_in_test(backend_config)
+
+    def check_multi_devices_forward(self, device_0, device_1):
+        layer, hook = self.layer, self.hook
+        layer.add_hook(hook)
+        layer.to_device(device_1)
+        x = device_1.send(self.x)
+
+        msg = None
+        with chainer.using_device(device_0):
+            try:
+                layer(x)
+            except Exception as e:
+                msg = e
+        assert msg is None
+
+    @attr.chainerx
+    @attr.multi_gpu(2)
+    def test_forward_chx_on_multi_devices(self):
+        if not self.lazy_init:
+            device_0 = BackendConfig(
+                {'use_chainerx': True, 'chainerx_device': 'cuda:0'}).device
+            device_1 = BackendConfig(
+                {'use_chainerx': True, 'chainerx_device': 'cuda:1'}).device
+            self.check_multi_devices_forward(device_0, device_1)
+
+    @attr.multi_gpu(2)
+    def test_forward_multi_gpus(self):
+        if not self.lazy_init:
+            device_0 = BackendConfig(
+                {'use_cuda': True, 'cuda_device': 0}).device
+            device_1 = BackendConfig(
+                {'use_cuda': True, 'cuda_device': 1}).device
+            self.check_multi_devices_forward(device_0, device_1)
+
+    def check_serialization(self, backend_config):
+        root = tempfile.mkdtemp()
+        filename = os.path.join(root, 'tmp.npz')
+
+        layer1 = self.layer.copy('copy')
+        hook1 = copy.deepcopy(self.hook)
+        layer1.add_hook(hook1)
+
+        layer1.to_device(backend_config.device)
+        x = backend_config.get_array(self.x)
+        with backend_config:
+            layer1(x)
+            with chainer.using_config('train', False):
+                y1 = layer1(x)
+        serializers.save_npz(filename, layer1)
+
+        layer2 = self.layer.copy('copy')
+        hook2 = copy.deepcopy(self.hook)
+        layer2.add_hook(hook2)
+
+        # Test loading is nice.
+        msg = None
+        try:
+            serializers.load_npz(filename, layer2)
+        except Exception as e:
+            msg = e
+        assert msg is None
+
+        with chainer.using_config('train', False):
+            y2 = layer2(self.x.copy())
+
+        # Test attributes are the same.
+        orig_weight = to_cpu(getattr(layer1, hook1.weight_name).array)
+        orig_vector = to_cpu(getattr(layer1, hook1.vector_name))
+        numpy.testing.assert_array_equal(
+            orig_weight, getattr(layer2, hook2.weight_name).array)
+        numpy.testing.assert_array_equal(
+            orig_vector, getattr(layer2, hook2.vector_name))
+        testing.assert_allclose(y1.array, y2.array, **self.allclose_options)
+
+        shutil.rmtree(root)
```

Comment:
Could you make sure the line is executed even if the test fails?  Do you think `chainer.utils.tempdir` can be used here?

Reply:
I didn't know the existence of `chainer.utils.tempdir`. Thank you.